File: leetcode/79_Word_Search.py
import unittest
from typing import List

class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        N, M = len(board), len(board[0])
        def rec(board: List[List[str]], used: List[List[bool]], word: str, i: int, j: int, idx: int):
            if idx >= len(word):
                return True
            for (next_i, next_j) in ((i + 1, j), (i, j + 1), (i - 1, j), (i , j - 1)):
                if  0 <= next_i < N and 0 <= next_j < M and word[idx] == board[next_i][next_j] and not used[next_i][next_j]:
                    used[next_i][next_j] = True
                    if rec(board, used, word, next_i, next_j, idx + 1):
                        return True
                    used[next_i][next_j] = False
            return False
            
        used = [[False] * M for i in range(N)]
        for i in range(N):
            for j in range(M):
                if word[0] == board[i][j]:
                    used[i][j] = True
                    if rec(board, used, word, i, j, 1):
                        return True
                    used[i][j] = False
        return False

class TestExist(unittest.TestCase):
    def test_1(self):
        self.assertTrue(Solution().exist([["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]], "ABCCED"))    

    def test_2(self):
        self.assertTrue(Solution().exist([["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]], "SEE"))    

    def test_3(self):
        self.assertFalse(Solution().exist([["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]], "ABCB"))    

    def test_3(self):
        self.assertTrue(Solution().exist([["C","A","A"],["A","A","A"],["B","C","D"]], "AAB"))    

File: leetcode/79_Word_Search.rs
struct Solution {}

impl Solution {
    pub fn exist(board: Vec<Vec<char>>, word: String) -> bool {
        const DIJ: [(i32, i32); 4] = [(0, 1), (1, 0), (0, -1), (-1, 0)];
        fn rec(board: &Vec<Vec<char>>, word: &Vec<char>, used: &mut Vec<Vec<bool>>, i: usize, j: usize, idx: usize) -> bool {
            let (n, m) = (board.len() as i32, board[0].len() as i32);
            if word[idx] == board[i][j] {
                if idx == word.len() - 1 {
                    return true;
                }
                used[i][j] = true;
                for (di, dj) in &DIJ {
                    let (next_i , next_j) = (i as i32 + di, j as i32 + dj);
                    if 0 <= next_i && next_i < n && 0 <= next_j && next_j < m && !used[next_i as usize][next_j as usize] {
                        if rec(board, word, used, next_i as usize, next_j as usize, idx + 1) {
                            return true;
                        }
                    }
                }
                used[i][j] = false;
            }
            false
        }
        let (n, m) = (board.len(), board[0].len());
        let word = word.chars().collect();
        let mut used = vec![vec![false; m]; n];
        for i in 0..n {
            for j in 0..m {
                if rec(&board, &word, &mut used, i, j, 0) {
                    return true;
                }
            }
        }
        false
    }
}

#[cfg(test)]
mod exist_test {
    use super::*;

    #[test]
    fn test_1() {
        assert!(Solution::exist(vec![vec!['A','B','C','E'], vec!['S','F','C','S'], vec!['A','D','E','E']], "ABCCED".to_string()));
    }    

    #[test]
    fn test_2() {
        assert!(Solution::exist(vec![vec!['A','B','C','E'], vec!['S','F','C','S'], vec!['A','D','E','E']], "SEE".to_string()));
    }    

    #[test]
    fn test_3() {
        assert!(!Solution::exist(vec![vec!['A','B','C','E'], vec!['S','F','C','S'], vec!['A','D','E','E']], "ABCB".to_string()));
    }    

    #[test]
    fn test_4() {
        assert!(Solution::exist(vec![vec!['C','A','A'], vec!['A','A','A'], vec!['B','C','D']], "AAB".to_string()));
    }    
}